File: CS/CustomizeRibbonTemplateExample.Win/DetailRibbonForm1.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.Templates;
using DevExpress.ExpressApp.Templates.ActionControls;
using DevExpress.ExpressApp.Win.Controls;
using DevExpress.ExpressApp.Win.SystemModule;
using DevExpress.ExpressApp.Win.Templates;
using DevExpress.ExpressApp.Win.Templates.Utils;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Ribbon;

namespace CustomizeRibbonTemplateExample.Win
{
    public partial class DetailRibbonForm1 : RibbonForm, IActionControlsSite, IContextMenuHolder, IWindowTemplate, IBarManagerHolder, ISupportViewChanged, ISupportUpdate, IViewSiteTemplate, ISupportStoreSettings, IViewHolder
    {
        private static readonly object viewChanged = new object();
        private static readonly object settingsReloaded = new object();
        private StatusMessagesHelper statusMessagesHelper;

        protected virtual void RaiseViewChanged(DevExpress.ExpressApp.View view)
        {
            EventHandler<TemplateViewChangedEventArgs> handler = (EventHandler<TemplateViewChangedEventArgs>)Events[viewChanged];
            if (handler != null)
            {
                handler(this, new TemplateViewChangedEventArgs(view));
            }
        }
        protected virtual void RaiseSettingsReloaded()
        {
            EventHandler handler = (EventHandler)Events[settingsReloaded];
            if (handler != null)
            {
                handler(this, EventArgs.Empty);
            }
        }

        public DetailRibbonForm1()
        {
            InitializeComponent();
            ribbonControl.Manager.ForceLinkCreate();
            statusMessagesHelper = new StatusMessagesHelper(barContainerStatusMessages);
        }

        #region IActionControlsSite Members
        IEnumerable<IActionControlContainer> IActionControlsSite.ActionContainers
        {
            get { return ribbonControl.ActionContainers; }
        }
        IEnumerable<IActionControl> IActionControlsSite.ActionControls
        {
            get { return ribbonControl.ActionControls; }
        }
        IActionControlContainer IActionControlsSite.DefaultContainer
        {
            get { return barActionContainerView; }
        }
        #endregion

        #region IFrameTemplate Members
        void IFrameTemplate.SetView(DevExpress.ExpressApp.View view)
        {
            viewSiteManager.SetView(view);
            RaiseViewChanged(view);
        }
        ICollection<IActionContainer> IFrameTemplate.GetContainers()
        {
            return new IActionContainer[] { };
        }
        IActionContainer IFrameTemplate.DefaultContainer
        {
            get { return null; }
        }
        #endregion

        #region IWindowTemplate Members
        void IWindowTemplate.SetCaption(string caption)
        {
            Text = caption;
        }
        void IWindowTemplate.SetStatus(ICollection<string> statusMessages)
        {
            statusMessagesHelper.SetMessages(statusMessages);
        }
        bool IWindowTemplate.IsSizeable
        {
            get { return FormBorderStyle == FormBorderStyle.Sizable; }
            set { FormBorderStyle = (value ? FormBorderStyle.Sizable : FormBorderStyle.FixedDialog); }
        }
        #endregion

        #region IBarManagerHolder Members
        BarManager IBarManagerHolder.BarManager
        {
            get { return ribbonControl.Manager; }
        }
        event EventHandler IBarManagerHolder.BarManagerChanged
        {
            add { }
            remove { }
        }
        #endregion

        #region IContextMenuHolder
        PopupMenu IContextMenuHolder.ContextMenu
        {
            get { return contextMenu; }
        }
        #endregion

        #region ISupportViewChanged Members
        event EventHandler<TemplateViewChangedEventArgs> ISupportViewChanged.ViewChanged
        {
            add { Events.AddHandler(viewChanged, value); }
            remove { Events.RemoveHandler(viewChanged, value); }
        }
        #endregion

        #region ISupportUpdate Members
        void ISupportUpdate.BeginUpdate()
        {
            if (ribbonControl.Manager != null)
            {
                ribbonControl.Manager.BeginUpdate();
            }
        }
        void ISupportUpdate.EndUpdate()
        {
            if (ribbonControl.Manager != null)
            {
                ribbonControl.Manager.EndUpdate();
            }
        }
        #endregion

        #region IViewSiteTemplate Members
        object IViewSiteTemplate.ViewSiteControl
        {
            get { return viewSiteManager.ViewSiteControl; }
        }
        #endregion

        #region ISupportStoreSettings Members
        void ISupportStoreSettings.SetSettings(IModelTemplate settings)
        {
            IModelTemplateWin templateModel = (IModelTemplateWin)settings;
            TemplatesHelper templatesHelper = new TemplatesHelper(templateModel);
            IModelFormState formState;
            if (viewSiteManager.View != null)
            {
                formState = templatesHelper.GetFormStateNode(viewSiteManager.View.Id);
            }
            else {
                formState = templatesHelper.GetFormStateNode();
            }
            formStateModelSynchronizer.Model = formState;
            templatesHelper.SetRibbonSettings(ribbonControl);
        }
        void ISupportStoreSettings.ReloadSettings()
        {
            modelSynchronizationManager.ApplyModel();
            RaiseSettingsReloaded();
        }
        void ISupportStoreSettings.SaveSettings()
        {
            SuspendLayout();
            try
            {
                modelSynchronizationManager.SynchronizeModel();
            }
            finally
            {
                ResumeLayout();
            }
        }
        event EventHandler ISupportStoreSettings.SettingsReloaded
        {
            add { Events.AddHandler(settingsReloaded, value); }
            remove { Events.RemoveHandler(settingsReloaded, value); }
        }
        #endregion

        #region IViewHolder Members
        DevExpress.ExpressApp.View IViewHolder.View
        {
            get { return viewSiteManager.View; }
        }
        #endregion
    }
}

File: VB/CustomizeTemplate.Module.Win/Custom Templates/DetailRibbonForm1.vb
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.Templates
Imports DevExpress.ExpressApp.Templates.ActionControls
Imports DevExpress.ExpressApp.Win.Controls
Imports DevExpress.ExpressApp.Win.SystemModule
Imports DevExpress.ExpressApp.Win.Templates
Imports DevExpress.ExpressApp.Win.Templates.Utils
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Ribbon

Partial Public Class DetailRibbonForm1
    Inherits RibbonForm
    Implements IActionControlsSite
    Implements IContextMenuHolder
    Implements IWindowTemplate
    Implements IBarManagerHolder
    Implements ISupportViewChanged
    Implements ISupportUpdate
    Implements IViewSiteTemplate
    Implements ISupportStoreSettings
    Implements IViewHolder
    Private Shared ReadOnly viewChanged As New Object()
    Private Shared ReadOnly settingsReloaded As New Object()
    Private statusMessagesHelper As StatusMessagesHelper

    Protected Overridable Sub RaiseViewChanged(view As DevExpress.ExpressApp.View)
        RaiseEvent ISupportViewChanged_ViewChanged(Me, New TemplateViewChangedEventArgs(view))
    End Sub
    Protected Overridable Sub RaiseSettingsReloaded()
        RaiseEvent ISupportStoreSettings_SettingsReloaded(Me, EventArgs.Empty)
    End Sub

    Public Sub New()
        InitializeComponent()
        ribbonControl.Manager.ForceLinkCreate()
        statusMessagesHelper = New StatusMessagesHelper(barContainerStatusMessages)
    End Sub

#Region "IActionControlsSite Members"
    Private ReadOnly Property IActionControlsSite_ActionContainers() As IEnumerable(Of IActionControlContainer) Implements IActionControlsSite.ActionContainers
        Get
            Return ribbonControl.ActionContainers
        End Get
    End Property
    Private ReadOnly Property IActionControlsSite_ActionControls() As IEnumerable(Of IActionControl) Implements IActionControlsSite.ActionControls
        Get
            Return ribbonControl.ActionControls
        End Get
    End Property
    Private ReadOnly Property IActionControlsSite_DefaultContainer() As IActionControlContainer Implements IActionControlsSite.DefaultContainer
        Get
            Return barActionContainerView
        End Get
    End Property
#End Region

#Region "IFrameTemplate Members"
    Private Sub IFrameTemplate_SetView(view As DevExpress.ExpressApp.View) Implements IFrameTemplate.SetView
        viewSiteManager.SetView(view)
        RaiseViewChanged(view)
    End Sub
    Private Function IFrameTemplate_GetContainers() As ICollection(Of IActionContainer) Implements IFrameTemplate.GetContainers
        Return New IActionContainer() {}
    End Function
    Private ReadOnly Property IFrameTemplate_DefaultContainer() As IActionContainer Implements IFrameTemplate.DefaultContainer
        Get
            Return Nothing
        End Get
    End Property
#End Region

#Region "IWindowTemplate Members"
    Private Sub IWindowTemplate_SetCaption(caption As String) Implements IWindowTemplate.SetCaption
        Text = caption
    End Sub
    Private Sub IWindowTemplate_SetStatus(statusMessages As ICollection(Of String)) Implements IWindowTemplate.SetStatus
        statusMessagesHelper.SetMessages(statusMessages)
    End Sub
    Private Property IWindowTemplate_IsSizeable() As Boolean Implements IWindowTemplate.IsSizeable
        Get
            Return FormBorderStyle.Sizable = FormBorderStyle
        End Get
        Set(value As Boolean)
            FormBorderStyle = If(value, FormBorderStyle.Sizable, FormBorderStyle.FixedDialog)
        End Set
    End Property
#End Region

#Region "IBarManagerHolder Members"
    Private ReadOnly Property IBarManagerHolder_BarManager() As BarManager Implements IBarManagerHolder.BarManager
        Get
            Return ribbonControl.Manager
        End Get
    End Property
    Private Custom Event IBarManagerHolder_BarManagerChanged As EventHandler Implements IBarManagerHolder.BarManagerChanged
        AddHandler(ByVal value As EventHandler)
        End AddHandler
        RemoveHandler(ByVal value As EventHandler)
        End RemoveHandler
        RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
        End RaiseEvent
    End Event
#End Region

#Region "IContextMenuHolder"
    Private ReadOnly Property IContextMenuHolder_ContextMenu() As PopupMenu Implements IContextMenuHolder.ContextMenu
        Get
            Return templateContextMenu
        End Get
    End Property
#End Region

#Region "ISupportViewChanged Members"
    Private Custom Event ISupportViewChanged_ViewChanged As EventHandler(Of TemplateViewChangedEventArgs) Implements ISupportViewChanged.ViewChanged
        AddHandler(ByVal value As EventHandler(Of TemplateViewChangedEventArgs))
            Events.AddHandler(viewChanged, value)
        End AddHandler
        RemoveHandler(ByVal value As EventHandler(Of TemplateViewChangedEventArgs))
            Events.RemoveHandler(viewChanged, value)
        End RemoveHandler
        RaiseEvent(ByVal sender As Object, ByVal e As TemplateViewChangedEventArgs)
            Dim handler As EventHandler(Of TemplateViewChangedEventArgs) = CType(Events(viewChanged), EventHandler(Of TemplateViewChangedEventArgs))
            If (handler IsNot Nothing) Then
                handler.Invoke(sender, e)
            End If
        End RaiseEvent
    End Event
#End Region

#Region "ISupportUpdate Members"
    Private Sub ISupportUpdate_BeginUpdate() Implements ISupportUpdate.BeginUpdate
        If ribbonControl.Manager IsNot Nothing Then
            ribbonControl.Manager.BeginUpdate()
        End If
    End Sub
    Private Sub ISupportUpdate_EndUpdate() Implements ISupportUpdate.EndUpdate
        If ribbonControl.Manager IsNot Nothing Then
            ribbonControl.Manager.EndUpdate()
        End If
    End Sub
#End Region

#Region "IViewSiteTemplate Members"
    Private ReadOnly Property IViewSiteTemplate_ViewSiteControl() As Object Implements IViewSiteTemplate.ViewSiteControl
        Get
            Return viewSiteManager.ViewSiteControl
        End Get
    End Property
#End Region

#Region "ISupportStoreSettings Members"
    Private Sub ISupportStoreSettings_SetSettings(settings As IModelTemplate) Implements ISupportStoreSettings.SetSettings
        Dim templateModel As IModelTemplateWin = DirectCast(settings, IModelTemplateWin)
        Dim templatesHelper As New TemplatesHelper(templateModel)
        Dim formState As IModelFormState
        If ViewSiteManager.View IsNot Nothing Then
            formState = templatesHelper.GetFormStateNode(ViewSiteManager.View.Id)
        Else
            formState = templatesHelper.GetFormStateNode()
        End If
        formStateModelSynchronizer.Model = formState
        templatesHelper.SetRibbonSettings(RibbonControl)
    End Sub
    Private Sub ISupportStoreSettings_ReloadSettings() Implements ISupportStoreSettings.ReloadSettings
        modelSynchronizationManager.ApplyModel()
        RaiseSettingsReloaded()
    End Sub
    Private Sub ISupportStoreSettings_SaveSettings() Implements ISupportStoreSettings.SaveSettings
        SuspendLayout()
        Try
            modelSynchronizationManager.SynchronizeModel()
        Finally
            ResumeLayout()
        End Try
    End Sub
    Private Custom Event ISupportStoreSettings_SettingsReloaded As EventHandler Implements ISupportStoreSettings.SettingsReloaded
        AddHandler(ByVal value As EventHandler)
            Events.AddHandler(settingsReloaded, value)
        End AddHandler
        RemoveHandler(ByVal value As EventHandler)
            Events.RemoveHandler(settingsReloaded, value)
        End RemoveHandler
        RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
            Dim handler As EventHandler = CType(Events(settingsReloaded), EventHandler)
            If (handler IsNot Nothing) Then
                handler.Invoke(sender, e)
            End If
        End RaiseEvent
    End Event
#End Region

#Region "IViewHolder Members"
    Private ReadOnly Property IViewHolder_View() As DevExpress.ExpressApp.View Implements IViewHolder.View
        Get
            Return viewSiteManager.View
        End Get
    End Property
#End Region
End Class
